diff --git a/Makefile.am b/Makefile.am
index a7a11ad..a8f30b2 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -89,14 +89,14 @@ memcached_SOURCES += tls.c tls.h
 endif
 
 memcached_debug_SOURCES = $(memcached_SOURCES)
-memcached_CPPFLAGS = -DNDEBUG
+memcached_CPPFLAGS = -DNDEBUG -I../../../Symlib/include -I../kcut
 memcached_debug_LDADD = @PROFILER_LDFLAGS@
 memcached_debug_CFLAGS = @PROFILER_FLAGS@
 
 memcached_LDADD =
 memcached_LDFLAGS =
 memcached_debug_LDFLAGS =
-memcached_DEPENDENCIES =
+memcached_DEPENDENCIES = ../kcut/libextension.a
 memcached_debug_DEPENDENCIES =
 CLEANFILES=
 
@@ -131,7 +131,7 @@ memcached_debug_LDADD += vendor/liburing/src/liburing.a
 endif
 
 memcached_debug_CFLAGS += -DMEMCACHED_DEBUG
-
+memcached_LDADD +=  -L ../kcut -lc -lextension  ../../../Symlib/build/L1/L1.a -lkallsyms
 # build fails on Darwin with const signature replacements.
 if DARWIN
 memcached_dtrace.h: memcached_dtrace.d
diff --git a/memcached.c b/memcached.c
index 1858d02..1215e3b 100644
--- a/memcached.c
+++ b/memcached.c
@@ -61,6 +61,9 @@
 #include <sys/sysctl.h>
 #endif
 
+#include "LINF/sym_all.h"
+#include "../kcut/kcut_tcpmsg.h"
+
 /*
  * forward declarations
  */
@@ -126,10 +129,13 @@ enum transmit_result {
     TRANSMIT_HARD_ERROR  /** Can't write (c->state is set to conn_closing) */
 };
 
+
 /* Default methods to read from/ write to a socket */
 ssize_t tcp_read(conn *c, void *buf, size_t count) {
     assert (c != NULL);
-    return read(c->sfd, buf, count);
+
+    int ret = kcut_tcp_recvmsg(c->sfd, buf, count);
+    return ret;
 }
 
 ssize_t tcp_sendmsg(conn *c, struct msghdr *msg, int flags) {
@@ -139,7 +145,9 @@ ssize_t tcp_sendmsg(conn *c, struct msghdr *msg, int flags) {
 
 ssize_t tcp_write(conn *c, void *buf, size_t count) {
     assert (c != NULL);
-    return write(c->sfd, buf, count);
+
+    int ret = kcut_tcp_sendmsg(c->sfd, buf, count);
+    return ret;
 }
 
 static enum transmit_result transmit(conn *c);
@@ -4692,6 +4700,8 @@ static int _mc_meta_load_cb(const char *tag, void *ctx, void *data) {
     return reuse_mmap;
 }
 
+
+ 
 int main (int argc, char **argv) {
     int c;
     bool lock_memory = false;
@@ -6010,6 +6020,7 @@ int main (int argc, char **argv) {
         settings.proxy_ctx = proxy_init(settings.proxy_uring, settings.proxy_memprofile);
     }
 #endif
+
 #ifdef EXTSTORE
     memcached_thread_init(settings.num_threads, storage);
     init_lru_crawler(storage);
@@ -6179,7 +6190,6 @@ int main (int argc, char **argv) {
 
     /* Initialize the uriencode lookup table. */
     uriencode_init();
-
     /* enter the event loop */
     while (!stop_main_loop) {
         if (event_base_loop(main_base, EVLOOP_ONCE) != 0) {
@@ -6187,7 +6197,6 @@ int main (int argc, char **argv) {
             break;
         }
     }
-
     switch (stop_main_loop) {
         case GRACE_STOP:
             fprintf(stderr, "Gracefully stopping\n");
diff --git a/thread.c b/thread.c
index 9ddbefe..0380c7d 100644
--- a/thread.c
+++ b/thread.c
@@ -501,6 +501,7 @@ static void setup_thread(LIBEVENT_THREAD *me) {
     thread_io_queue_add(me, IO_QUEUE_NONE, NULL, NULL);
 }
 
+#include "LINF/sym_all.h"
 /*
  * Worker thread: main event loop
  */
@@ -521,6 +522,10 @@ static void *worker_libevent(void *arg) {
     }
 
     register_thread_initialized();
+#ifndef BRACKET_PRIV
+    sym_elevate();
+#endif
+
     while (!event_base_got_exit(me->base)) {
         event_base_loop(me->base, EVLOOP_ONCE);
         // Run IO queues after the event loop to catch things like
@@ -532,6 +537,9 @@ static void *worker_libevent(void *arg) {
         }
 #endif
     }
+#ifndef BRACKET_PRIV
+    symbi_fast_lower();
+#endif
     // same mechanism used to watch for all threads exiting.
     register_thread_initialized();
 
