diff --git a/Makefile.am b/Makefile.am
index a7a11ad..7435b28 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -88,10 +88,13 @@ if ENABLE_TLS
 memcached_SOURCES += tls.c tls.h
 endif
 
+KCUT_FINAL_FLAGS = $(KCUT_CFLAGS)
+
 memcached_debug_SOURCES = $(memcached_SOURCES)
-memcached_CPPFLAGS = -DNDEBUG
+memcached_CPPFLAGS = -DNDEBUG $(KCUT_FINAL_FLAGS)
 memcached_debug_LDADD = @PROFILER_LDFLAGS@
-memcached_debug_CFLAGS = @PROFILER_FLAGS@
+memcached_debug_CFLAGS = @PROFILER_FLAGS@ $(KCUT_FINAL_FLAGS)
+
 
 memcached_LDADD =
 memcached_LDFLAGS =
@@ -131,6 +134,8 @@ memcached_debug_LDADD += vendor/liburing/src/liburing.a
 endif
 
 memcached_debug_CFLAGS += -DMEMCACHED_DEBUG
+memcached_debug_LDADD += $(KCUT_LDFLAGS)
+memcached_LDADD += $(KCUT_LDFLAGS)
 
 # build fails on Darwin with const signature replacements.
 if DARWIN
@@ -214,6 +219,15 @@ endif
 	  done \
 	else :; fi
 
+memcached$(EXEEXT): $(memcached_OBJECTS) $(memcached_DEPENDENCIES) $(EXTRA_memcached_DEPENDENCIES) 
+	@rm -f memcached$(EXEEXT)
+	$(AM_V_CCLD)$(memcached_LINK) $(memcached_OBJECTS) $(memcached_LDADD) $(LIBS)
+	${KLD} ${KLDFLAGS} -o $@
+
+memcached-debug$(EXEEXT): $(memcached_debug_OBJECTS) $(memcached_debug_DEPENDENCIES) $(EXTRA_memcached_debug_DEPENDENCIES) 
+	@rm -f memcached-debug$(EXEEXT)
+	$(AM_V_CCLD)$(memcached_debug_LINK) $(memcached_debug_OBJECTS) $(memcached_debug_LDADD) $(LIBS)
+	${KLD} ${KLDFLAGS} -o $@
 docs:
 	(cat Doxyfile ; echo "PROJECT_NUMBER=`cat version.num`") | doxygen -
 
diff --git a/memcached.c b/memcached.c
index 1858d02..308c589 100644
--- a/memcached.c
+++ b/memcached.c
@@ -61,6 +61,8 @@
 #include <sys/sysctl.h>
 #endif
 
+#include "kcut_tcpmsg.h"
+
 /*
  * forward declarations
  */
@@ -126,10 +128,13 @@ enum transmit_result {
     TRANSMIT_HARD_ERROR  /** Can't write (c->state is set to conn_closing) */
 };
 
+
 /* Default methods to read from/ write to a socket */
 ssize_t tcp_read(conn *c, void *buf, size_t count) {
     assert (c != NULL);
-    return read(c->sfd, buf, count);
+
+    int ret = kcut_tcp_read(c->sfd, buf, count);
+    return ret;
 }
 
 ssize_t tcp_sendmsg(conn *c, struct msghdr *msg, int flags) {
@@ -139,7 +144,9 @@ ssize_t tcp_sendmsg(conn *c, struct msghdr *msg, int flags) {
 
 ssize_t tcp_write(conn *c, void *buf, size_t count) {
     assert (c != NULL);
-    return write(c->sfd, buf, count);
+
+    int ret = kcut_tcp_write(c->sfd, buf, count);
+    return ret;
 }
 
 static enum transmit_result transmit(conn *c);
@@ -4692,6 +4699,8 @@ static int _mc_meta_load_cb(const char *tag, void *ctx, void *data) {
     return reuse_mmap;
 }
 
+
+ 
 int main (int argc, char **argv) {
     int c;
     bool lock_memory = false;
@@ -6010,6 +6019,7 @@ int main (int argc, char **argv) {
         settings.proxy_ctx = proxy_init(settings.proxy_uring, settings.proxy_memprofile);
     }
 #endif
+
 #ifdef EXTSTORE
     memcached_thread_init(settings.num_threads, storage);
     init_lru_crawler(storage);
@@ -6179,7 +6189,6 @@ int main (int argc, char **argv) {
 
     /* Initialize the uriencode lookup table. */
     uriencode_init();
-
     /* enter the event loop */
     while (!stop_main_loop) {
         if (event_base_loop(main_base, EVLOOP_ONCE) != 0) {
@@ -6187,7 +6196,6 @@ int main (int argc, char **argv) {
             break;
         }
     }
-
     switch (stop_main_loop) {
         case GRACE_STOP:
             fprintf(stderr, "Gracefully stopping\n");
diff --git a/thread.c b/thread.c
index 9ddbefe..b3f07ca 100644
--- a/thread.c
+++ b/thread.c
@@ -501,6 +501,7 @@ static void setup_thread(LIBEVENT_THREAD *me) {
     thread_io_queue_add(me, IO_QUEUE_NONE, NULL, NULL);
 }
 
+#include "kcut.h"
 /*
  * Worker thread: main event loop
  */
@@ -521,6 +522,9 @@ static void *worker_libevent(void *arg) {
     }
 
     register_thread_initialized();
+
+    kcut_init();
+
     while (!event_base_got_exit(me->base)) {
         event_base_loop(me->base, EVLOOP_ONCE);
         // Run IO queues after the event loop to catch things like
@@ -532,6 +536,9 @@ static void *worker_libevent(void *arg) {
         }
 #endif
     }
+    
+    kcut_cleanup();
+
     // same mechanism used to watch for all threads exiting.
     register_thread_initialized();
 
