diff --git a/Makefile.am b/Makefile.am
index a7a11ad..8290128 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -89,14 +89,14 @@ memcached_SOURCES += tls.c tls.h
 endif
 
 memcached_debug_SOURCES = $(memcached_SOURCES)
-memcached_CPPFLAGS = -DNDEBUG
+memcached_CPPFLAGS = -DNDEBUG -I../../../Symlib/include -I../kcut
 memcached_debug_LDADD = @PROFILER_LDFLAGS@
 memcached_debug_CFLAGS = @PROFILER_FLAGS@
 
 memcached_LDADD =
 memcached_LDFLAGS =
 memcached_debug_LDFLAGS =
-memcached_DEPENDENCIES =
+memcached_DEPENDENCIES = ../kcut/libextension.a
 memcached_debug_DEPENDENCIES =
 CLEANFILES=
 
@@ -131,7 +131,7 @@ memcached_debug_LDADD += vendor/liburing/src/liburing.a
 endif
 
 memcached_debug_CFLAGS += -DMEMCACHED_DEBUG
-
+memcached_LDADD +=  -L ../kcut -lc -lextension -lkallsyms -L ../../../Symlib/dynam_build -lSym
 # build fails on Darwin with const signature replacements.
 if DARWIN
 memcached_dtrace.h: memcached_dtrace.d
diff --git a/memcached.c b/memcached.c
index 1858d02..0062df7 100644
--- a/memcached.c
+++ b/memcached.c
@@ -61,6 +61,11 @@
 #include <sys/sysctl.h>
 #endif
 
+#include "kernel.h"
+#include "LINF/sym_all.h"
+#include "L1/stack_switch.h"
+#include "../kcut/kcut_tcpmsg.kh"
+
 /*
  * forward declarations
  */
@@ -126,10 +131,79 @@ enum transmit_result {
     TRANSMIT_HARD_ERROR  /** Can't write (c->state is set to conn_closing) */
 };
 
+#ifdef BRACKET_STACK
+#include <dlfcn.h>
+
+__thread unsigned long kcut_ktos = 0;
+
+static inline int
+resolve_sym(char *name, void **value)
+{
+  char *error;
+  dlerror();  // clear as per manpage
+  *value = dlsym(RTLD_DEFAULT, name);
+  error = dlerror();
+  if (error != NULL) {
+    fprintf(stderr, "%s\n", error);
+    return 0;
+  }
+  return 1;
+}
+
+static inline void set_kcut_ktos(void)
+{
+  if (!resolve_sym("cpu_current_top_of_stack", (void **)&kcut_ktos)) {
+    printf("failed to resolve cpu_current_top_of_stack\n");
+    assert(0);
+  }
+}
+#endif
+
 /* Default methods to read from/ write to a socket */
 ssize_t tcp_read(conn *c, void *buf, size_t count) {
     assert (c != NULL);
-    return read(c->sfd, buf, count);
+
+    struct iovec iov;
+    iov.iov_base = buf;
+    iov.iov_len = count;
+
+#ifdef BRACKET_STACK
+    if (!kcut_ktos) set_kcut_ktos();
+#endif
+
+    const int KCUT_THRESHOLD = 20;
+    static int kcut_cnt = 0;
+    int ret;    
+    if (kcut_cnt) {
+#ifdef BRACKET_PRIV
+#ifdef BRACKET_STACK
+        // privilege switching and stack switching
+        sym_elevate();
+        SYM_ON_KERN_STACK_DYNSYM_DO_CONST_PRIV(kcut_ktos, ret = kcut_tcp_recvmsg(c->sfd, &iov));
+        symbi_fast_lower();
+#else
+        // privilege switch only, stay on user stack
+        sym_elevate();
+        ret = kcut_tcp_recvmsg(c->sfd, &iov);
+        symbi_fast_lower();
+#endif 
+#else // no privilege switch, assume we have already elevated
+#ifdef BRACKET_STACK
+        // no privilege switch, switch to kernel stack
+        SYM_ON_KERN_STACK_DYNSYM_DO_CONST_PRIV(kcut_ktos, ret = kcut_tcp_recvmsg(c->sfd, &iov));
+#else
+        // no privilege switch, remain on user stack
+        ret = kcut_tcp_recvmsg(c->sfd, &iov);
+#endif
+#endif // BRACKET_PRIV
+
+        kcut_cnt--;
+    } else {
+        ret = read(c->sfd, buf, count);
+        kcut_cnt = KCUT_THRESHOLD;
+    }
+
+    return ret;
 }
 
 ssize_t tcp_sendmsg(conn *c, struct msghdr *msg, int flags) {
@@ -139,7 +213,48 @@ ssize_t tcp_sendmsg(conn *c, struct msghdr *msg, int flags) {
 
 ssize_t tcp_write(conn *c, void *buf, size_t count) {
     assert (c != NULL);
-    return write(c->sfd, buf, count);
+
+    struct iovec iov;
+    iov.iov_base = buf;
+    iov.iov_len = count;
+
+#ifdef BRACKET_STACK
+    if (!kcut_ktos) set_kcut_ktos();
+#endif
+
+    const int KCUT_THRESHOLD = 20;
+    static int kcut_cnt = 0;
+    int ret;    
+    if (kcut_cnt) {
+#ifdef BRACKET_PRIV
+#ifdef BRACKET_STACK
+        // privilege switching and stack switching
+        sym_elevate();
+        SYM_ON_KERN_STACK_DYNSYM_DO_CONST_PRIV(kcut_ktos, ret = kcut_tcp_sendmsg(c->sfd, &iov));
+        symbi_fast_lower();
+#else
+        // privilege switch only, stay on user stack
+        sym_elevate();
+        ret = kcut_tcp_sendmsg(c->sfd, &iov);
+        symbi_fast_lower();
+#endif 
+#else // no privilege switch, assume we have already elevated
+#ifdef BRACKET_STACK
+        // no privilege switch, switch to kernel stack
+        SYM_ON_KERN_STACK_DYNSYM_DO_CONST_PRIV(kcut_ktos, ret = kcut_tcp_sendmsg(c->sfd, &iov));
+#else
+        // no privilege switch, remain on user stack
+        ret = kcut_tcp_sendmsg(c->sfd, &iov);
+#endif
+#endif // BRACKET_PRIV
+
+        kcut_cnt--;
+    } else {
+        ret = write(c->sfd, buf, count);
+        kcut_cnt = KCUT_THRESHOLD;
+    }
+
+    return ret;
 }
 
 static enum transmit_result transmit(conn *c);
@@ -4692,6 +4807,8 @@ static int _mc_meta_load_cb(const char *tag, void *ctx, void *data) {
     return reuse_mmap;
 }
 
+
+ 
 int main (int argc, char **argv) {
     int c;
     bool lock_memory = false;
@@ -6179,7 +6296,11 @@ int main (int argc, char **argv) {
 
     /* Initialize the uriencode lookup table. */
     uriencode_init();
-
+    
+#ifndef BRACKET_PRIV
+    sym_elevate();
+#endif
+    
     /* enter the event loop */
     while (!stop_main_loop) {
         if (event_base_loop(main_base, EVLOOP_ONCE) != 0) {
@@ -6187,7 +6308,10 @@ int main (int argc, char **argv) {
             break;
         }
     }
-
+#ifndef BRACKET_PRIV
+    symbi_fast_lower();
+#endif
+    
     switch (stop_main_loop) {
         case GRACE_STOP:
             fprintf(stderr, "Gracefully stopping\n");
